recipes/full_finetune_distributed.py

```diff
@@ -614,6 +671,60 @@ def _setup_data(
 
         return sampler, dataloader
 
+    def save_checkpoint_async(
+        self,
+        epoch: int,
+    ) -> None:
+        """
+        Checkpoint the state of the recipe. The constructed checkpoint state dict
+        contains the following information:
+        - Model weights with key training.MODEL_KEY
+        - Relevant recipe state if training is not complete
+
+        Checkpointer will save the model weights and recipe state in an asynchrnous way
+        which unblocks the training sooner to continue. To correctly resume training from
+        an intermediate checkpoint, user needs to have both resume_from_checkpoint and
+        enable_async_checkpointing flags set to True in the config.
+        """
+
+        if self._is_rank_zero:
+            log.info("Saving checkpoint asynchronously. Retrieving full state dict...")
+            cp_start = time.perf_counter()
+
+        # Create the checkpoint dict to be sent to the checkpointer and ultimately persisted to storage
+        ckpt_dict = {
+            training.SEED_KEY: self.seed,
+            training.EPOCHS_KEY: self.epochs_run,
+            training.TOTAL_EPOCHS_KEY: self.total_epochs,
+            training.MAX_STEPS_KEY: self.max_steps_per_epoch,
+        }
+
+        model_state_dict = self._model.state_dict()
+        optim_state_dict = (
+            self._optim_ckpt_wrapper.state_dict()
+            if self._optimizer_in_bwd
+            else self._optimizer.state_dict()
+        )
+
+        ckpt_dict[training.MODEL_KEY] = model_state_dict
+        ckpt_dict[training.OPT_KEY] = optim_state_dict
+
+        dcp_saver = DistributedCheckpointer(
+            checkpoint_dir=self._checkpointer._checkpoint_dir,
+            output_dir=self._checkpointer._output_dir,
+        )
```

Comment:
Just wanna check my understanding here. We normally would define checkpointer as part of the config to be instantiated but here you basically use `enable_async_checkpointing` as a proxy for whether we're the `DistributedCheckpointer` or not. Is the idea behind this to have the ability to save and load intermediate checkpoints with `DistributedCheckpointer` but still save the final checkpoint in the HF/Meta format?

Reply:
Yes. Exactly. I was less inclined to have users configure two different Checkpointers in the same job config: `DistributedCheckpointer` for intermediate and HF/Meta checkpointer for first load and final save. I exposed the async checkpointing as a feature instead under the `enable_async_checkpointing` config for two reasons:
- `DistributedCheckpointer` is the only checkpointer which supports async checkpointing. 
- Users will probably want to control the first load and final save checkpoints. Intermediate checkpoints are for fault tolerance and we have flexibility to enforce the Checkpointer that works best for that.

Basically the end state I was thinking of is the following:
- For first load and final save, users can choose from any of the following Checkpointers: HF, Meta or `DistributedCheckpointer`. No async checkpointing here, since checkpoint loads and final saves will always need to be synchronous. We can not finish the job early until the entire final checkpoint has been persisted. 
- For intermediate checkpoints, we use async save with `DistributedCheckpointer` by default.